File: iOS/mobileflux/mobileflux/middleware/Logger.swift
//
//  Logger.swift
//  mobileflux
//
//  Created by RORY KELLY on 23/09/2020.
//

import Foundation
import RxSwift

protocol Logger {
    func log(loggable: String)
}

protocol Loggable {
    func log() -> String
}


func withLogger<S: State>(logger: Logger?) -> (Store<S>) -> Completable {
    return { store in
        return store.actions
            .do(onNext: { it in logger?.log(loggable: "dispatched: \(it) state: \(store.currentViewState()) ")})
                .ignoreElements()
    }
}
       
    

File: android/rxflux-core/src/main/java/com/klefmusic/rxfluxcore/middleware/Logger.kt
package com.klefmusic.rxfluxcore.middleware

import com.klefmusic.rxfluxcore.ReduxStore
import com.klefmusic.rxfluxcore.withMiddleware
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.CompletableSource
import io.reactivex.rxjava3.kotlin.ofType

interface Logger {

    fun log(loggable: String)

}

interface Loggable {
    fun log(): String = javaClass.name
}

fun <S> ReduxStore<S>.withLogger(logger: Logger?): ReduxStore<S> =
    withMiddleware { store ->
        val ignoreElements: CompletableSource =
            store.actions.doOnNext { logger?.log("dispatched: $it state: ${store.currentViewState()} ") }
                .ofType()
                .ignoreElements()
        Completable.mergeArray(
            ignoreElements
        )
    }